core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala

```diff
@@ -45,83 +45,125 @@ class SpendReportingService(userInfo: UserInfo, dataSource: SlickDataSource, big
   }
 
-  private def extractWorkspaceSpendAggregation(rows: List[FieldValueList], currency: Currency, startTime: DateTime, endTime: DateTime, workspaceProjectsToNames: Map[GoogleProject, WorkspaceName]): SpendReportingAggregation = {
-    val workspaceSpend = rows.map { row =>
-      val rowCost = BigDecimal(row.get("cost").getDoubleValue).setScale(currency.getDefaultFractionDigits, RoundingMode.HALF_EVEN)
-      val rowCredits = BigDecimal(row.get("credits").getDoubleValue).setScale(currency.getDefaultFractionDigits, RoundingMode.HALF_EVEN)
-      val rowGoogleProjectId = GoogleProject(row.get("googleProjectId").getStringValue)
-      val rowWorkspaceName = workspaceProjectsToNames.getOrElse(rowGoogleProjectId, throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadGateway, s"unexpected project ${rowGoogleProjectId.value} returned by BigQuery")))
-      SpendReportingForDateRange(rowCost.toString(),
-        rowCredits.toString(),
-        currency.getCurrencyCode,
-        startTime,
-        endTime,
-        Option(rowWorkspaceName),
-        Option(rowGoogleProjectId))
-    }
-    SpendReportingAggregation(
-      SpendReportingAggregationKeys.Workspace, workspaceSpend
-    )
+  /**
+    * Ensure that BigQuery results only include one type of currency and return that currency.
+    */
+  private def getCurrency(rows: List[FieldValueList]): Currency = {
+    val currencies = rows.map(_.get("currency").getStringValue)
+
+    Currency.getInstance(currencies.reduce { (x, y) =>
+      if (x.equals(y)) {
+        x
+      } else {
+        throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadGateway, s"Inconsistent currencies found while aggregating spend data: $x and $y cannot be combined"))
+      }
+    })
   }
 
-  private def extractDailySpendAggregation(rows: List[FieldValueList], currency: Currency): SpendReportingAggregation = {
-    val dailySpend = rows.map { row =>
-      val rowCost = BigDecimal(row.get("cost").getDoubleValue).setScale(currency.getDefaultFractionDigits, RoundingMode.HALF_EVEN)
-      val rowCredits = BigDecimal(row.get("credits").getDoubleValue).setScale(currency.getDefaultFractionDigits, RoundingMode.HALF_EVEN)
-      SpendReportingForDateRange(rowCost.toString(),
-        rowCredits.toString(),
-        currency.getCurrencyCode,
-        DateTime.parse(row.get("date").getStringValue),
-        DateTime.parse(row.get("date").getStringValue).plusDays(1).minusMillis(1))
+  private def extractSpendAggregation(rows: List[FieldValueList], currency: Currency, aggregationKey: SpendReportingAggregationKeyWithSub, workspaceProjectsToNames: Map[GoogleProject, WorkspaceName] = Map.empty): SpendReportingAggregation = {
+    aggregationKey match {
+      case SpendReportingAggregationKeyWithSub(SpendReportingAggregationKeys.Category, subAggregationKey) => extractCategorySpendAggregation(rows, currency, subAggregationKey, workspaceProjectsToNames)
```

Comment:
What happens if the subaggregation key is the same as the aggregation key? Does the code handle that gracefully ? I can imagine that either the subaggregation key is ignored, or an error is thrown.

Reply:
It does handle it gracefully, although the report that's returned isn't very useful since the data in the sub-aggregation has already been aggregated in the same say. Here's a snippet of the response

```
        {
          "category": "Other",
          "cost": "0.07",
          "credits": "0.00",
          "currency": "USD",
          "subAggregation": {
            "aggregationKey": "Category",
            "spendData": [
              {
                "category": "Other",
                "cost": "0.07",
                "credits": "0.00",
                "currency": "USD"
              }
            ]
          }
        }